### ISSUE TITLE
Report ambiguous attribute keys earlier than in `expression.Value()`

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/hcl#648]
## Background

Currently ambiguous attribute keys are only detected when/if the expression is being evaluated:

https://github.com/hashicorp/hcl/blob/c964a71ca32006c9e7a0730b5a0e1dd60b05d308/hclsyntax/expression.go#L1293-L1298

In the context of the language server, we don't evaluate expressions in the sense that we don't care about the values. Terraform (core) likely evaluates them at last minute, to ensure that they get evaluated in the correct order.

## Proposal

I am not sure if there is yet a precedent for diagnostics concerning expressions being returned from  [`hclsyntax.ParseConfig()`](https://pkg.go.dev/github.com/hashicorp/hcl/v2@v2.19.1/hclsyntax#ParseConfig) but I wanted to suggest for that function to return some equivalent diagnostic.

For what it's worth the language server could relatively easily produce the diagnostic on its own, so this isn't a deal breaker. However, the benefit of the parser producing it is that all future language servers can agree on how that diagnostic is produced.

To add a counter-argument here, we can also achieve some unity by producing that diagnostic inside of `hcl-lang` and still avoid touching `hcl`.

TL;DR this is more of an exploratory design question than a must have.

--- 

Context: https://github.com/hashicorp/terraform-ls/issues/1550

**Comment by apparentlymart** [comment]
If I recall correctly, this works the way it does because we wanted this rule to apply only to "normal" expression evaluation, and not to other situations like passing the expression to `hcl.ExprMap`.

In that sort of direct static analysis, the caller just gets the raw key expression and decides for themselves how to use it. For example, in the `providers` argument in `module` blocks in Terraform, the "keys" are provider addresses rather than normal expressions, and so there's no such ambiguity in that location and thus no need to return this error. (and it would be a breaking change to do so)

With all of that said then, I think this must stay roughly how it is to retain the flexibility of evaluation that Terraform (at least) depends on.

**Comment by radeksimko** [comment]
That is completely fair. I appreciate the explanation, thank you!

Given that it's so contextual then I think it's also even more appropriate that we handle it outside of HCL, where we can deal with that context.
